_posts/2025-04-11-Hybrid-Search-Options-And-Practice.md

```diff
@@ -0,0 +1,295 @@
+              "k": 5
+            }
+          }
+        }
+      ]
+    }
+  }
+}
+```
+
+This query uses  the following parameters:
+
+- `hybrid`: OpenSearch's built-in hybrid query type
+- `search_pipeline`: Applies the min-max normalization and weighted score combination
+
+### How min-max normalization works
+
+Min-max normalization works as follows:
+
+1. OpenSearch runs both the `match` and `neural` queries.
+1. The pipeline normalizes each result set using min-max scaling.
+1. Normalized scores are combined using a weighted average to produce the final score.
+
+In this example, the final document scores are calculated as:  
+
+```r
+0.3 * normalized_BM25_score + 0.7 * normalized_neural_score
+```
+
+## When to use score-based normalization
+
+Score normalization is useful in scenarios where you want fine-tuned control over how results from different methods are combined. Consider using it when:
+
+- **You can calibrate or train weights**: If you have a reliable way to normalize scores (for example, using min-max or L2 norms) and **tune the weight of each retrieval method**, a linear score combination can slightly outperform rank fusion in relevance metrics.
+- **You want to emphasize one method**: Score normalization allows you to explicitly emphasize the more reliable model for your use case. In a general document search, some queries might be answered mostly by keyword matching while others need semantic understanding. By adjusting weights (or using query-dependent logic), you can give BM25 or the neural model more influence when appropriate. 
+- **You need precise control over merging (fewer false positives)**: A normalized linear combination gives you more control over how results merge, which can reduce the chance of a less relevant document being boosted solely because of one method. For example, if a semantic model returns a somewhat irrelevant result at a high rank, rank-based normalization (described in the next section) still makes this result high-scoring because of its rank. A score-based approach can mitigate this by yielding a lower normalized relevance score for that document (because this document might be scored at 0 by BM25, and the semantic score alone may not be enough to beat other combined scores). Thus, in a well-tuned hybrid system, score normalization can lead to a more precise top-10 ranking, because each document's final score reflects a balanced lexical and semantic relevance.
+
+
+## Hybrid search using RRF
+
+Rank-based techniques focus on a document's position in the result set, rather than its raw score. These techniques are especially useful when combining results from different retrieval methods whose scores aren't directly comparable. 
+
+**RRF** is one such technique. It uses the rank of each document in individual query results to calculate a combined score, making it robust against mismatched scoring scales across methods like BM25 and semantic search.
+
```

Comment:
Line 198: I'm not following the meaning of "robust against". Please replace with clearer phrasing.

Reply:
Reworded.